# 4.8 Task - Grouping Data & Aggregating Variables 

## This script contains the following points:

#### 1. Import libraries and dataframe
#### 2. Reducing the data format
#### 3. Aggregated mean for entire dataframe
#### 4. Aggregating Data with transform()
#### 5. Deriving Columns with loc()
#### 6. Create a spending flag
#### 7. Create an order frequency flag

#### 1. Import libraries and dataframe

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [4]:
# Defining path and import dataframe
path=r'C:\Users\olilo\OneDrive\Dokumente\CF\Data Analytics Immersion_Week 8-12\2023-08-28 Instacart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_new_columns.pkl'))

In [5]:
# Output check
ords_prods_merge.shape

(32404859, 20)

#### 2. Reducing the data format

In [6]:
# Information about data size
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 20 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   eval_set                object  
 3   order_number            int64   
 4   order_dow               int64   
 5   order_hour_of_day       int64   
 6   days_since_prior_order  float64 
 7   product_id              int64   
 8   add_to_cart_order       int64   
 9   reordered               int64   
 10  product_name            object  
 11  aisle_id                int64   
 12  department_id           int64   
 13  prices                  float64 
 14  _merge                  category
 15   price_range_loc        object  
 16  price_range_loc         object  
 17  busiest_day             object  
 18  busiest_days            object  
 19  busiest_period_of_day   object  
dtypes: category(1), float64(2), int64(10), objec

In [7]:
# Checking the information
ords_prods_merge.tail(5)

order_id  user_id eval_set  order_number  order_dow  \
32404854   1320836   202557    prior            17          2   
32404855     31526   202557    prior            18          5   
32404856    758936   203436    prior             1          2   
32404857   2745165   203436    prior             2          3   
32404858   3093936   205420    prior             1          4   

          order_hour_of_day  days_since_prior_order  product_id  \
32404854                 15                     1.0       43553   
32404855                 11                     3.0       43553   
32404856                  7                     NaN       42338   
32404857                  5                    15.0       42338   
32404858                 14                     NaN       28818   

          add_to_cart_order  reordered                   product_name  \
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices _merge  price_range_loc  \
32404854        64              7     3.7   both              NaN   
32404855        64              7     3.7   both              NaN   
32404856        50             19     6.9   both              NaN   
32404857        50             19     6.9   both              NaN   
32404858       130             14    10.3   both              NaN   

            price_range_loc     busiest_day         busiest_days  \
32404854  Low-range product  Regularly busy  Regularly busy days   
32404855  Low-range product  Regularly busy  Regularly busy days   
32404856  Mid-range product  Regularly busy  Regularly busy days   
32404857  Mid-range product  Regularly busy      Least busy days   
32404858  Mid-range product      Least busy      Least busy days   

         busiest_period_of_day  
32404854           Most orders  
32404855           Most orders  
32404856        Average orders  
32404857         Fewest orders  
32404858           Most orders

In [8]:
#Reducing the data format
ords_prods_merge['order_number'] = ords_prods_merge['order_number'].astype('uint8')

In [9]:
ords_prods_merge['order_dow'] = ords_prods_merge['order_dow'].astype('uint8')

In [10]:
ords_prods_merge['order_hour_of_day'] = ords_prods_merge['order_hour_of_day'].astype('uint8')

In [11]:
ords_prods_merge['reordered'] = ords_prods_merge['reordered'].astype('uint8')

In [12]:
ords_prods_merge['department_id'] = ords_prods_merge['department_id'].astype('uint8')

In [13]:
ords_prods_merge['product_id'] = ords_prods_merge['product_id'].astype('uint32')

In [14]:
# Information about data size after reduction
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 20 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   eval_set                object  
 3   order_number            uint8   
 4   order_dow               uint8   
 5   order_hour_of_day       uint8   
 6   days_since_prior_order  float64 
 7   product_id              uint32  
 8   add_to_cart_order       int64   
 9   reordered               uint8   
 10  product_name            object  
 11  aisle_id                int64   
 12  department_id           uint8   
 13  prices                  float64 
 14  _merge                  category
 15   price_range_loc        object  
 16  price_range_loc         object  
 17  busiest_day             object  
 18  busiest_days            object  
 19  busiest_period_of_day   object  
dtypes: category(1), float64(2), int64(4), object

#### 3. Aggregated mean for entire dataframe

In [15]:
# Aggregated mean of the “order_number” column grouped by “department_id” for whole dataframe
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [16]:
# Tests for proving a significant difference between the means of different groups
# Using the mean() function to achieve the same results 
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

Results from df.groupby('department_id')['order_number'].mean(), as a sample:
department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

The values differ from each other in some cases only slightly, such as in the department_id 4 sample (18.825780), and department_id 4 ( 17.811403) of the entire dataframe. Comparing that the sample had a million columns, the full dataframe has 32.404.859 the difference betwwwn means is small.

#### 4. Aggregating Data with transform()

In [17]:
# Split the data into groups
# Apply the transform() function
# Create a new column,'max_order'
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [18]:
# Check out the results
ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   2968173       15    prior            15          1                  9   
11   1870022       15    prior            17          2                 16   
12   1911383       15    prior            18          2                 11   
13   2715276       15    prior            21          1                  9   
14    487368       15    prior            22          1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                      NaN         196                  1          0  ...   
1                     15.0         196                  1          1  ...   
2                     21.0         196                  1          1  ...   
3                     29.0         196                  1          1  ...   
4                     28.0         196                  1          1  ...   
5                     19.0         196                  1          1  ...   
6                     20.0         196                  1          1  ...   
7                     14.0         196                  2          1  ...   
8                      0.0         196                  4          1  ...   
9                     30.0         196                  1          1  ...   
10                     7.0         196                  2          0  ...   
11                     8.0         196                  6          1  ...   
12                     7.0         196                  1          1  ...   
13                     7.0         196                  1          1  ...   
14                    14.0         196                  1          1  ...   

   aisle_id  department_id  prices  _merge  price_range_loc  \
0        77              7     9.0    both              NaN   
1        77              7     9.0    both              NaN   
2        77              7     9.0    both              NaN   
3        77              7     9.0    both              NaN   
4        77              7     9.0    both              NaN   
5        77              7     9.0    both              NaN   
6        77              7     9.0    both              NaN   
7        77              7     9.0    both              NaN   
8        77              7     9.0    both              NaN   
9        77              7     9.0    both              NaN   
10       77              7     9.0    both              NaN   
11       77              7     9.0    both              NaN   
12       77              7     9.0    both              NaN   
13       77              7     9.0    both              NaN   
14       77              7     9.0    both              NaN   

      price_range_loc     busiest_day         busiest_days  \
0   Mid-range product  Regularly busy  Regularly busy days   
1   Mid-range product  Regularly busy      Least busy days   
2   Mid-range product  Regularly busy      Least busy days   
3   Mid-range product      Least busy      Least busy days   
4   Mid-range product      Least busy      Least busy days   
5   Mid-range product  Regularly busy  Regularly busy days   
6   Mid-range product  Regularly busy         Busiest days   
7   Mid-range pro

In [19]:
# Command to assign any options regarding the maximum number of rows
pd.options.display.max_rows = None

In [20]:
# Checking output again
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   2968173       15    prior            15          1                  9   
11   1870022       15    prior            17          2                 16   
12   1911383       15    prior            18          2                 11   
13   2715276       15    prior            21          1                  9   
14    487368       15    prior            22          1                 10   
15   2293453       19    prior             2          5                 14   
16   1973799       19    prior             5          6                 12   
17    532817       19    prior             7          4                 17   
18   1573906       21    prior            10          3                 10   
19   1593000       31    prior            10          3                  8   
20   2231262       31    prior            17          3                 11   
21   2580647       43    prior             6          4                 16   
22   2187180       43    prior             9          4                 12   
23   2497897       52    prior             1          1                  9   
24   1318871       52    prior             2          1                 10   
25   1261384       52    prior             3          1                 10   
26    580568       52    prior             5          2                 10   
27   2428073       52    prior             7          2                  7   
28    342306       52    prior             9          1                 11   
29   1520353       52    prior            10          2                 10   
30   2542086       52    prior            11          2                  9   
31    944694       52    prior            12          1                 16   
32   1268191       52    prior            15          2                 11   
33   1122089       52    prior            17          2                 10   
34   1498922       52    prior            18          2                 10   
35    180919       52    prior            19          3                  8   
36   2695875       52    prior            21          5                 10   
37   3205960       67    prior             1          3                 11   
38    970922       67    prior             2          1                 15   
39    596585       67    prior             3          1                 13   
40   2037087       67    prior             4          2                  9   
41   2474111       67    prior             5          3                 14   
42   1568481       67    prior             6          3                 12   
43   2048527       67    prior             8          2                 11   
44   2389031       67    prior             9          2                 14   
45   2378562       67    prior            10          1                 10   
46   1515698       67    prior            11          1                 11   
47   1928050       67    prior            12          2                 12   
48   1949274       67    prior            13          1                 10   
49    551869       67    prior            14          1                  9   
50    768572       67    p

#### 5. Deriving Columns with loc()

In [21]:
# Creating a new column, Loyal custimer
# Flag with loc function
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [22]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [23]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [24]:
# Value check, count conclusion
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [25]:
# Check multiple columns at the same time with new synta
# loyalty flag + check the basic statistics of the product prices for each loyalty category
print('The mean, min and max of prices grouped by the loyalty_flag column in df_orders_products_merged:')
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by the loyalty_flag column in df_orders_products_merged:


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

No visible difference in the comparison of the three groups regarding the lowest and highest value spent. The Average of New customers is higher than the Average from Loyal customers group, This is surprising, but at the same time it translates into a potential increase in loyal customers.

#### 6. Create a spending flag

In [26]:
# Group by prices
ords_prods_merge.groupby('prices')

In [27]:
# Aggregated mean of the “loyalty_flag” column grouped by “prices” for whole dataframe
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

In [28]:
# Grouping the dataframe by 'user_id' and 'prices'
ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [29]:
#  Output check
ords_prods_merge.shape

(32404859, 23)

In [30]:
ords_prods_merge.head(20)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   2968173       15    prior            15          1                  9   
11   1870022       15    prior            17          2                 16   
12   1911383       15    prior            18          2                 11   
13   2715276       15    prior            21          1                  9   
14    487368       15    prior            22          1                 10   
15   2293453       19    prior             2          5                 14   
16   1973799       19    prior             5          6                 12   
17    532817       19    prior             7          4                 17   
18   1573906       21    prior            10          3                 10   
19   1593000       31    prior            10          3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                      NaN         196                  1          0  ...   
1                     15.0         196                  1          1  ...   
2                     21.0         196                  1          1  ...   
3                     29.0         196                  1          1  ...   
4                     28.0         196                  1          1  ...   
5                     19.0         196                  1          1  ...   
6                     20.0         196                  1          1  ...   
7                     14.0         196                  2          1  ...   
8                      0.0         196                  4          1  ...   
9                     30.0         196                  1          1  ...   
10                     7.0         196                  2          0  ...   
11                     8.0         196                  6          1  ...   
12                     7.0         196                  1          1  ...   
13                     7.0         196                  1          1  ...   
14                    14.0         196                  1          1  ...   
15                     6.0         196                  3          0  ...   
16                     8.0         196                 15          1  ...   
17                     6.0         196                  1          1  ...   
18                     6.0         196                  2          0  ...   
19                     7.0         196                 17          0  ...   

   prices  _merge   price_range_loc    price_range_loc     busiest_day  \
0     9.0    both               NaN  Mid-range product  Regularly busy   
1     9.0    both               NaN  Mid-range product  Regularly busy   
2     9.0    both               NaN  Mid-range product  Regularly busy   
3     9.0    both               NaN  Mid-range product      Least busy   
4     9.0    both               NaN  Mid-range product      Least busy   
5     9.0    both               NaN  Mid-range product  Regularly busy   
6     9.0    both               NaN  Mid-range product  Regularly busy   
7     9.0    both               NaN  Mid-range product  Regularly busy   
8     9.0    both               NaN  Mid-range product  Regularly busy   
9     9

In [31]:
# Creating a new column, Average_price
# Spending Flag with loc function
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'Spending_flag'] = 'Low spender'

In [32]:
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'Spending_flag'] = 'High spender'

In [33]:
# Value check, count conclusion
ords_prods_merge['Spending_flag'].value_counts(dropna = False)

Spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

#### 7. Create an order frequency flag

In [34]:
# Grouping the dataframe by 'user_id' and 'days_since_prior_order'
ords_prods_merge['median_days_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [35]:
#  Output check
ords_prods_merge.shape

(32404859, 25)

In [36]:
ords_prods_merge.head(20)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   2968173       15    prior            15          1                  9   
11   1870022       15    prior            17          2                 16   
12   1911383       15    prior            18          2                 11   
13   2715276       15    prior            21          1                  9   
14    487368       15    prior            22          1                 10   
15   2293453       19    prior             2          5                 14   
16   1973799       19    prior             5          6                 12   
17    532817       19    prior             7          4                 17   
18   1573906       21    prior            10          3                 10   
19   1593000       31    prior            10          3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                      NaN         196                  1          0  ...   
1                     15.0         196                  1          1  ...   
2                     21.0         196                  1          1  ...   
3                     29.0         196                  1          1  ...   
4                     28.0         196                  1          1  ...   
5                     19.0         196                  1          1  ...   
6                     20.0         196                  1          1  ...   
7                     14.0         196                  2          1  ...   
8                      0.0         196                  4          1  ...   
9                     30.0         196                  1          1  ...   
10                     7.0         196                  2          0  ...   
11                     8.0         196                  6          1  ...   
12                     7.0         196                  1          1  ...   
13                     7.0         196                  1          1  ...   
14                    14.0         196                  1          1  ...   
15                     6.0         196                  3          0  ...   
16                     8.0         196                 15          1  ...   
17                     6.0         196                  1          1  ...   
18                     6.0         196                  2          0  ...   
19                     7.0         196                 17          0  ...   

    price_range_loc    price_range_loc     busiest_day         busiest_days  \
0               NaN  Mid-range product  Regularly busy  Regularly busy days   
1               NaN  Mid-range product  Regularly busy      Least busy days   
2               NaN  Mid-range product  Regularly busy      Least busy days   
3               NaN  Mid-range product      Least busy      Least busy days   
4               NaN  Mid-range product      Least busy      Least busy days   
5               NaN  Mid-range product  Regularly busy  Regularly busy days   
6               NaN  Mid-range product  Regularly busy         Busiest days   
7               NaN  Mid-range product  Regularly busy         Busiest days   
8               NaN  Mid-range produ

In [37]:
# Frequency Flag with loc function in three criteria
ords_prods_merge.loc[ords_prods_merge['median_days_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [38]:
ords_prods_merge.loc[(ords_prods_merge['median_days_order'] > 10) & (ords_prods_merge['median_days_order'] <= 20), 'frequency_flag'] = 'Regular customer'

In [39]:
ords_prods_merge.loc[ords_prods_merge['median_days_order'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [40]:
# Value check, count conclusion
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [41]:
# Print Not a Number, NaN values
print(ords_prods_merge[ords_prods_merge.frequency_flag.isnull()])

          order_id  user_id eval_set  order_number  order_dow  \
13645692    895835   159838    prior             1          0   
17251990    895835   159838    prior             1          0   
17622767    895835   159838    prior             1          0   
24138593    895835   159838    prior             1          0   
25880002    895835   159838    prior             1          0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered  ...    price_range_loc  busiest_day  \
13645692                  3          0  ...  Mid-range product  Busiest day   
17251990                  6          0  ...  Mid

These are 5 new customers, they do not have average orders (median_days_order), but they can be matched in the 'frequency_flag' column to Non-frequent customer.

In [42]:
# Correcting the NaN values in 'frequency_flag' column
ords_prods_merge['frequency_flag'].fillna("Non-frequent customer", inplace = True)

In [43]:
# Secound check, count conclusion
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636442
Name: count, dtype: int64

In [44]:
# Export data to pkl
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_agg.pkl'))